src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/HttpHandlerDiagnosticListener.cs

```diff
@@ -0,0 +1,711 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections;
+using System.Collections.Generic;
+using System.Net;
+using System.Reflection;
+using System.Reflection.Emit;
+using System.Runtime.Serialization;
+
+// This HttpHandlerDiagnosticListener class is applicable only for .NET 4.6, and not for .NET core.
+// If you are making these changes, please test your changes manually via custom test applications.
+
+namespace System.Diagnostics
+{
+    /// <summary>
+    /// A HttpHandlerDiagnosticListener is a DiagnosticListener for .NET 4.6 where HttpClient
+    /// doesn't have a DiagnosticListener built in. This class is not used for .NET Core because
+    /// HttpClient in .NET Core already emits DiagnosticSource events. This class compensates for
+    /// that in .NET 4.6. HttpHandlerDiagnosticListener has no public constructor. To use this, 
+    /// the application just needs to call <see cref="DiagnosticListener.AllListeners" /> and
+    /// <see cref="DiagnosticListener.AllListenerObservable.Subscribe(IObserver{DiagnosticListener})"/>,
+    /// then in the <see cref="IObserver{DiagnosticListener}.OnNext(DiagnosticListener)"/> method,
+    /// when it seems the System.Net.Http.Desktop source, subscribe to it. This will trigger the
+    /// initialization of this DiagnosticListener.
+    /// </summary>
+    internal sealed class HttpHandlerDiagnosticListener : DiagnosticListener
+    {
+        /// <summary>
+        /// Overriding base class implementation just to give us a chance to initialize.  
+        /// </summary>
+        public override IDisposable Subscribe(IObserver<KeyValuePair<string, object>> observer, Predicate<string> isEnabled)
+        {
+            Initialize();
+            return base.Subscribe(observer, isEnabled);
+        }
+
+        /// <summary>
+        /// Overriding base class implementation just to give us a chance to initialize.  
+        /// </summary>
+        public override IDisposable Subscribe(IObserver<KeyValuePair<string, object>> observer, Func<string, object, object, bool> isEnabled)
+        {
+            Initialize();
+            return base.Subscribe(observer, isEnabled);
+        }
+
+        /// <summary>
+        /// Initializes all the reflection objects it will ever need. Reflection is costly, but it's better to take
+        /// this one time performance hit than to get it multiple times later, or do it lazily and have to worry about
+        /// threading issues. If Initialize has been called before, it will not doing anything.
+        /// </summary>
+        private void Initialize()
+        {
+            lock (this)
```

Comment:
Suggest: Use double-checked locking and handle the exception outside the lock. Like this:
```C#
private static Exception PrepareReflection()
{
    try
    {
        PrepareReflectionObjects(); 
        PerformInjection(); 
        return null;
    }
    catch (Exception ex)
    {
        return ex;
    }
}

private void Initialize()
{
    if (this.initialized)
    {
        return;
    }

    Exception ex;

    lock (this)
    {
        if (this.initialized)
        {
            return;
        }

        // This flag makes sure we only do this once. Even if we failed to initialize in an 
        // earlier time, we should not retry because this initialization is not cheap and 
        // the likelihood it will succeed the second time is very small. 
        this.initialized = true; 
        ex = PrepareReflection(); 
    }

    // If anything went wrong, just no-op. Write an event so at least we can find out. 
    if (ex != null) 
    {
        this.Write(InitializationFailed, new { Exception = ex });
    }
}